### PR TITLE
`nimdigger`: build nim at any revision since v0.12.0~157,  1 liner `git bisect` to help find regressions

### LINKED ISSUE #332: nim-digger: program to build nim at any revision (including forks/PRs) (cf D's digger); useful for git bisect
```
git bisect start badrev goodrev
git bisect run zsh -c "source ~/.zshrc && nim=$nim_0190_X d_nim_build_simple_local --skipusercfg && $nimb c -r --skipusercfg --skipparentcfg $timn_D/tests/nim/all/t11021.nim"
```

## note
* this requires knowing which csources hash was used to build a particular commit
* future csources2 should have a commit hash version controlled in nim repo so we can programmatically access it (EDIT: indeed, see nimBuildCsourcesIfNeeded in 1.5.1)

## EDIT
* for new nim versions, use `. ci/funs.sh && nimBuildCsourcesIfNeeded` (which also already handles caching)
* for older nim versions relying on csources, figure out which csources hash to checkout based on `git fetch --tags` and `git tag` and finding the latest tag, the using a table lookup to identify the appropriate csources hash. Sadly necessary because of lack of prior design but not very complicated (followup here for this: https://github.com/timotheecour/Nim/issues/332)

## links
in D: https://github.com/CyberShadow/Digger
```
Digger can:
build and test D from git
build older versions of D
build D plus forks and pending pull requests
bisect D's history to find where regressions are introduced (or bugs fixed)
```

## links
* made easier after https://github.com/nim-lang/Nim/pull/17815
* https://github.com/dom96/choosenim/issues/256#issuecomment-848064479

### LINKED ISSUE #702: `./koch bisect`
## links
* [bisect bot · Issue #8 · juancarlospaco/nimbug](https://github.com/juancarlospaco/nimbug/issues/8)
* [nim-digger: program to build nim at any revision (including forks/PRs) (cf D's digger); useful for git bisect · Issue #332 · timotheecour/Nim](https://github.com/timotheecour/Nim/issues/332)
* requires https://github.com/timotheecour/Nim/issues/332

### PR DESCRIPTION
**Pull request by timotheecour**
`nimdigger` is a tool to:
* build nim at any revision (including custom branches), taking
care of details such as figuring out automatically the correct csources/csources_v1 revision to use.
* allow 1-liner git bisect to find nim regressions, take care of building nim (if not cached) at intermediate revisions 

## design goals
* ease of use: 1 liner for running `git bisect` workflows, or to build nim at past revisions
* performance: via caching both csources built binaries, and intermediate nim binaries
* lazyness: build artifacts on demand
* go as far back as possible, currently oldest buildable nim version is v0.12.0~157
* isolated: all artifacts are kept under a (configurable) `$HOME/.nimdigger/cache/Nim`

## examples
build at any revision >= v0.12.0~157
```
$ nim r tools/nimdigger.nim --compileNim --rev:v0.15.2~10
$ $HOME/.nimdigger/cache/Nim/bin/nim -v
Nim Compiler Version 0.15.2 (2021-05-28) [MacOSX: amd64] [...]
```
find a which commit introduced a regression
```
$ nim r tools/nimdigger.nim --oldnew:v0.19.0..v0.20.0 --bisectCmd:'bin/nim -v | grep 0.19.0'
66c0f7c3fb214485ca6cfd799af6e50798fcdf6d is the first REGRESSION commit
```
find a which commit introduced a bugfix
```
$ nim r tools/nimdigger.nim --oldnew:v0.19.0..v0.20.0 --bisectBugfix --bisectCmd:'bin/nim -v | grep 0.20.0'
be9c38d2659496f918fb39e129b9b5b055eafd88 is the first BUGFIX commit
```

## actual examples
for regression https://github.com/nim-lang/Nim/issues/16376:
copy this snippet to /tmp/t16376.nim:
```nim
type Matrix[T] = object
  data: T
proc randMatrix*[T](m, n: int, max: T): Matrix[T] = discard
proc randMatrix*[T](m, n: int, x: Slice[T]): Matrix[T] = discard
template randMatrix*[T](m, n: int): Matrix[T] = randMatrix[T](m, n, T(1.0))
let B = randMatrix[float32](20, 10)
```
```
$ nim r tools/nimdigger.nim --oldnew:v0.19.0..v0.20.0 -- bin/nim c --hints:off --skipparentcfg --skipusercfg /tmp/t16376.nim
fd16875561634e3ef24072631cf85eeead6213f2 is the first REGRESSION commit
```

for regression https://github.com/nim-lang/Nim/issues/16496:
```
$ nim r tools/nimdigger.nim --oldnew:v1.2.0..v1.4.0 -- bin/nim c -r -d:case2 --hints:off --skipparentcfg --skipusercfg /tmp/t16376.nim
8b66412a8bfebcd53f09d0e608b418a05d13516a is the first REGRESSION commit
```
for regression https://github.com/nim-lang/Nim/issues/18113:
```
$ nim r tools/nimdigger.nim --oldnew:v1.2.0..origin/devel -- bin/nim c -r -d:case3 --hints:off --skipparentcfg --skipusercfg /tmp/t16376.nim
e5873b3a9300897443f0f2e2db128e3463089003 is the first REGRESSION commit
```

## speed
the more you use the tool, the faster it gets because binaries get cached, eg re-running this takes 10 seconds:
```
nim r tools/nimdigger.nim --oldnew:v1.2.0..origin/devel -- bin/nim c -r -d:case3 --hints:off --skipparentcfg --skipusercfg /tmp/t16376.nim
```

## links
* fixes https://github.com/timotheecour/Nim/issues/332
* fixes https://github.com/timotheecour/Nim/issues/702
* addresses 1st part of https://github.com/juancarlospaco/nimbug/issues/8
* analog to D's https://github.com/CyberShadow/Digger

## future work
- [ ] enable running a command at specified set of past revisions (eg a whole range) to plot performance benchmarks across time (refs: https://github.com/timotheecour/Nim/issues/425)
- [ ] nimdustmite (refs https://github.com/nim-lang/Nim/issues/8276) for bug minimization, analog to D's excellent https://github.com/CyberShadow/DustMite
- [ ] machine readable output: `--dump:output.json` (refs https://github.com/nim-lang/Nim/pull/18119#discussion_r641869871)

### CODE REVIEW

File: tools/nimdigger.nim
Comment (juancarlospaco) [inline tools/nimdigger.nim:64]: After the merge this documentation is visible publicly on places like [TheIndex](https://nim-lang.github.io/Nim/theindex.html) ?, that would be great.

Would be nice to mention on the docs "this tool calls Git, Make, ... ".

File: tools/nimdigger.nim
Comment (juancarlospaco) [inline tools/nimdigger.nim:324]: Maybe would be great to be the first tool to implement the idea of having a `--dump`,
that outputs machine readable useful information as pretty printed JSON, 
then the same design can be implemented on other tools later, see
https://github.com/nim-lang/Nim/issues/13740#issuecomment-603481849

Comment (timotheecour) [reply tools/nimdigger.nim:324]: yes, I'm all for `--dump:output.json`, but this can be done as followup work, and is not hard with jsonutils, I've added it as future work on top

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim:64]: will do; @juancarlospaco can you please try out this PR and see if you find any bugs/usability issues/questions etc?

Comment (juancarlospaco) [reply tools/nimdigger.nim:64]: Yes.

Comment (timotheecour) [reply tools/nimdigger.nim:64]: > After the merge this documentation is visible publicly on places like TheIndex ?, that would be great.

done

> Would be nice to mention on the docs "this tool calls Git, Make, ... ".

done

File: tools/kochdocs.nim
Comment (Varriount) [inline tools/kochdocs.nim]: Why are these commented out?

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]: Is it possible to give these fields a better name?

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]: This would probably be better as a regular sequence of string literals (or at least spread onto multiple lines).

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]: `url` should probably be shell quoted as well.

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim:171]: Could `fmt` be used here instead? It's a bit more readable than `&`.

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]: This should attempt to use `XDG_DATA_DIR` too.

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]: I feel that this is unnecessary - the amount of typing it saves is minimal compared to the decreased readability and increased mental parsing required. Also, it's only used once in the entire file.

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim:217]: Does this handle Windows systems?

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]: Rather than manual string parsing and comparison, it might be better to use regular expressions or pegs (this applies to the other string comparison logic in the file too).

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]:
```suggestion
proc parseKeyVal(s: string): OrderedTable[string, string] =
  ## parse bash-like entries of the form key=val
  for index, line in splitLines(s):
    if line.len == 0 or line.startsWith "#":
      continue

    let item = split(line, "=", maxsplit = 1)
    doAssert item.len == 2, $(line, item)

    result[item[0]] = item[1]
```

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim]: Why does this have a '2' suffix?

**Review by Varriount — CHANGES_REQUESTED** [review]
This looks like a really useful tool! The source code could use some polishing though.

In particular, I think some identifiers could be changed to make the code more readable, and certain functions could be split up to increase code organization.

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim:171]: no because of `\n` (see strformat docs)

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim:217]: i don't have windows; can you please try?

File: tools/nimdigger.nim
Comment (juancarlospaco) [reply tools/nimdigger.nim:171]: Would be nice to deprecate `&`, the `&` thats duplicate for `fmt` not the `&` to concatenate. :thinking:

File: tools/kochdocs.nim
Comment (timotheecour) [reply tools/kochdocs.nim]: debug leftover, fixed

File: tools/nimdigger.nim
Comment (Varriount) [inline tools/nimdigger.nim:311]: This is not cross-platform. Git should be able to run the command without using a shell.

**Comment by Varriount** [comment]
Some other thoughts on this:
  - I think the logic handling cloning the csources and Nim repositories is unnecessary. Users will likely have these repositories already cloned.
  - Cache data should be stored in a temporary directory (in order to check):
    - the directory in the XDG_CACHE_HOME environment variable,
    - /tmp on *nix, or MacOS.
    - %TEMP% on Windows (I think there's getTemporaryDirectory function somewhere?).
  - Is recompiling the executable necessary?

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: changed

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: this doesn't compile, try it; that's because of iterator vs proc overload

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: done

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: a seq wouldn't add to readability, but i've spread over 2 lines

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: done

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: I'm also using this in https://github.com/nim-lang/fusion/pull/32 and intend to add it to std/sugar in subsequent PR which will refactor its different uses.

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: I thought about this but IMO the correct way to handle this would be via a cross-platform addition to std/os, say ` getDataDir()` (we already have `getConfigDir()` for `XDG_CONFIG_HOME`, but no such equivalent for `XDG_DATA_DIR`). Once this is added, nimdigger can reuse it instead. => https://github.com/timotheecour/Nim/issues/744

Note also that this follows the same pattern as `~/.nimble/nimblecache`

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: it would be shorter with my own parser library, but doesn't looker simpler/shorter with std/pegs or std/re or nre; feel free to propose if you have some working code that's simpler than what I wrote here though.

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim:311]: can't windows subsystem for linux work here? (as done with azure-pipelines.yml on windows) Otherwise, I think this can be addressed in followup PR by people more familiar with windows (unless you have a concrete suggestion that makes it work for windows)

**Comment by timotheecour** [comment]
> I think the logic handling cloning the csources and Nim repositories is unnecessary. Users will likely have these repositories already cloned.

It's necessary; an important design goal for nimdigger is isolation; it allows checking out revisions/bisecting without impacting your own git repo; it's much cleaner and safer to have nimdigger manage its own Nim/csources/csources_v1 repo (that can be wiped if needed) rather than reuse a users Nim checkout (which typically won't care about csources anymore since newer nim versions rely on csources_v1, not csources). Less risk of user data corruption etc.


> Cache data should be stored in a temporary directory (in order to check):

as I mentioned in https://github.com/nim-lang/Nim/pull/18119#discussion_r641914802, this should be handled instead as an API added to `std/os` which will offer a platform-independent way to access XDG_CACHE_HOME (as we already do with XDG_DATA_HOME via getConfigDir and TMP via getTempDir), instead of having tools like nimdigger redefine it. Once std/os adds this API, nimdigger will reuse it (separation of concerns). => https://github.com/timotheecour/Nim/issues/744

> Is recompiling the executable necessary?

since https://github.com/nim-lang/Nim/pull/17829, `nim r` caches binaries so it doesn't actually recompile; of course, `nim c --outdir:bin tools/nimdigger` is always possible

**Review by dom96 — CHANGES_REQUESTED** [review]
Is there a reason this needs to be in the Nim repo? I would push back against including it here and instead having it as its own standalone Nimble package.

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim:217]: note that azure-pipelines.yml runs on windows and calls nimBuildCsourcesIfNeeded which calls this, so with the right setup, this should work on windows; a windows user will have to confirm though (and future work by a windows user can improve windows support)

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: I've now simplified this code

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim:171]: read https://nim-lang.github.io/Nim/strformat.html#nimfmt-vsdot-nimamp :)

File: tools/nimdigger.nim
Comment (Araq) [reply tools/nimdigger.nim]: `strscans` can be used here and can be merged with `parseNimGitTag`

File: tools/nimdigger.nim
Comment (Araq) [reply tools/nimdigger.nim]: The name should be `sugar.construct` IMO.

**Comment by Araq** [comment]
I will test it on Windows. Not sure whether it should be a Nimble package or not, but it probably should. Nim core developers can be expected to type `nimble install nimdigger` or `koch nimdigger`, it should be mentioned in Nim's internal documentation.

**Comment by juancarlospaco** [comment]
Fusion ?. 🤔

**Comment by Varriount** [comment]
After some thought, I do agree with Araq that this should be a separate, installable utility.

File: tools/nimdigger.nim
Comment (Varriount) [reply tools/nimdigger.nim]: I still feel that the tradeoff between writeability and readability limits its usefulness though. In the cases where it would be useful (types with only a few fields) the fact that it's a macro significantly limits its usefulness - the reader must mentally adjust their expectations when they see the macro being used. In cases where there are types with many fields, it's almost always a better idea to explicitly show which values are being used for each field, as the reader is less likely to know the order of the fields.

File: tools/nimdigger.nim
Comment (Varriount) [reply tools/nimdigger.nim:217]: This will work *if* the user is using WSL (which they may not be, if they are using Nim to build Windows executables).

**Comment by dom96** [comment]
@juancarlospaco why? It's trivial to install via Nimble and if for some reason it's not then we should fix Nimble.

I think we should close this, this utility doesn't need to be in the same repo as the compiler.

**Comment by timotheecour** [comment]
I'm ok with a separate nimble package and linking to it in internal docs, I'll first address remaining comments and then update this PR accordingly

**Comment by narimiran** [comment]
+1 for this being a standalone nimble package.

File: tools/nimdigger.nim
Comment (Araq) [reply tools/nimdigger.nim]: I don't understand this remark but we can discuss this in an RFC about how we should have `sugar.construct`.

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: thanks, done

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: renamed to construct, RFC left TODO

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim]: now using instead `getEnv(NimDiggerEnv, getCacheDir("nimdigger"))` now that i've added os.getCacheDir

File: tools/nimdigger.nim
Comment (timotheecour) [reply tools/nimdigger.nim:217]: so please try on windows and report back with how to improve windows support :)

**Comment by timotheecour** [comment]
@araq done with all comments, now working on moving this to a separate repo; 

how about fusion/tools/nimdigger? the goal being to avoid multiplying repos (fewer repos has benefits eg code reuse, less boilerplate wrt testing, doc publishing, more potential contributors etc); note that fusion/tools/nimdigger is intended to also be usable as a library

this would then be runnable via:
```
nim r /pathto/fusion/tools/nimdigger.nim --compileNim --rev:v0.15.2~10
# in future work, `nim r pkg/fusion/tools/nimdigger` could be supported,
# allowing passing canonical module paths instead of having to pass actual paths)
```
or if `nimble build` accepts building multiple binaries (does it?) then it could also be usable via `nimdigger --compileNim --rev:v0.15.2~10`

this assumes ppl are ok with allowing fusion to have some modules with these dependencies:
* nim devel (instead of 1.0)
* cligen

if not, then i'll just make a separate nimble package for simplicity

## note
I expect other similar tools to be added in future, eg nimdustmite (https://github.com/nim-lang/Nim/issues/8276)

**Comment by Araq** [comment]
> if not, then i'll just make a separate nimble package for simplicity

Separate Nimble package please but feel free to name the nimble package in preparation for much more code/utilities of yours. We both agree on the value mono-repos.

**Pull request closed without merging**
